scripts/lint-style.lean

```diff
@@ -4,21 +4,72 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Michael Rothgang
 -/
 
+import Batteries.Data.String.Basic
 import Mathlib.Tactic.Linter.TextBased
 import Cli.Basic
 
 /-!
 # Text-based style linters
 
-This files defines the `lint-style` executable which runs all text-based style linters.
+This file defines the `lint-style` executable which runs all text-based style linters.
 The linters themselves are defined in `Mathlib.Tactic.Linter.TextBased`.
 
-In addition, this checks that every file in `scripts` is documented in its top-level README.
+In addition, this checks that
+- `Mathlib.Init` is (transitively) imported in all of mathlib, and
+- every file in `scripts` is documented in its top-level README.
 -/
 
-open Cli Mathlib.Linter.TextBased
+open Cli Mathlib.Linter.TextBased System.FilePath
+
+/-- Check that `Mathlib.Init` is transitively imported in all of Mathlib.
+
+Every file imported in `Mathlib.Init` should in turn import the `Header` linter
+(except for the header linter itself, of course).
+Return `true` iff there was an error.
+-/
+def checkInitImports : IO Bool := do
+  -- Find any file in the Mathlib directory which does not contain any Mathlib import.
+  -- The current check (for any line starting with "import Mathlib") is not perfect
+  -- (for instance, it would detect a multi-line comment containing the line import Mathlib),
+  -- but is good enough in practice.
+  let mut modulesWithoutMathlibImports := #[]
+  let mut importsHeaderLinter := #[]
+  -- We simply parse `Mathlib.lean`; this removes the need for external tools like grep.
+  let allModules := (← IO.FS.lines "Mathlib.lean").map (·.stripPrefix "import ")
+    |>.filter (·.startsWith "Mathlib")
+  for module in allModules do
+    let path := (System.mkFilePath (module.split (· == '.'))).addExtension "lean"
+    let hasMathlibImport := (← IO.FS.lines path).any fun s ↦ s.startsWith "import Mathlib"
```

Comment:
Have you tried to use [docs#Lean.parseImports'](https://leanprover-community.github.io/mathlib4_docs/Lean/Elab/ParseImportsFast.html#Lean.parseImports')?  It promises to be fast...

Reply:
I hadn't realized that existed - this is indeed nicer. I've changed the code to use it, here and further below.

Reply:
Testing reveals that this function always seems to add a spurious module `Init`. Is this related to the Lean prelude? Anyway, this is easy to remove manually.